ckan/public-midnight-blue/base/javascript/toast.js

```diff
@@ -0,0 +1,192 @@
+   *    position: "bottom-right"
+   *    stacking: false
+   * });
+   *
+   *
+   * Options:
+   * - message   (string)  [required] The main message body of the toast.
+   * - type      (string)  Optional.  Defines toast style, matches keys from `styles`. Default: "default".
+   * - title     (string)  Optional.  Text shown in the toast header.
+   * - icon      (string)  Optional.  HTML for an icon displayed before the title.
+   * - subtitle  (string)  Optional.  Text shown in the header's small subtitle area.
+   * - delay     (number)  Optional.  Time in milliseconds before auto-hide. Default: 3000.
+   * - position  (string)  Optional.  Position key controlling toast placement. Default: "bottom-right".
+   * - stacking  (boolean) Optional.  Whether to stack toasts on top of each other. Default: true.
+  */
+  var toast = function (options) {
+    if (!options || !options.message) {
+      console.error("Toast: Missing required 'message' option.");
+      return;
+    }
+
+    const opts = {
+      type: "default",
+      title: "",
+      icon: "",
+      subtitle: "",
+      delay: 3000,
+      position: "bottom-right",
+      showProgress: true,
+      stacking: true,
+      ...options
+    };
+
+    const style = toast.styles[opts.type] || toast.styles.default;
+    const containerEl = toast._createToastContainer(opts);
+    const toastEl = document.createElement("div");
+    const toastID = `toast-${++toast.count}`;
+
+    toastEl.setAttribute("id", toastID);
+    toastEl.setAttribute("role", "alert");
+    toastEl.setAttribute("aria-live", "assertive");
+    toastEl.setAttribute("aria-atomic", "true");
+    toastEl.classList.add("toast", "align-items-center");
+    style.border && toastEl.classList.add(style.border);
+
+    toastEl.innerHTML = `
+            <div class="toast-header ${style.main}">
+                ${opts.icon}
+                <strong class="me-auto">${opts.title}</strong>
+                <small>${opts.subtitle}</small>
+                <button type="button" class="btn-close ${style.btnClose}" data-bs-dismiss="toast" aria-label="Close"></button>
+            </div>
+            <div class="toast-body position-relative">${opts.message}</div>
+        `;
+
+    if (!opts.stacking) {
+      containerEl.querySelectorAll(".toast").forEach(el => el.remove());
+    }
+
+    containerEl.appendChild(toastEl);
+
+    toastEl.addEventListener("hidden.bs.toast", (e) => e.target.remove());
+
+    const hasDelay = typeof opts.delay === "number" && opts.delay > 0;
+    const toastInstance = new bootstrap.Toast(toastEl, {
+      autohide: hasDelay,
+      delay: hasDelay ? opts.delay : 0
+    });
+
+    if (hasDelay && opts.showProgress) {
+      const progressEl = toast._createProgressBar(opts, style);
+
+      // reset progress bar to 100% on mouse enter
+      toastEl.addEventListener("mouseenter", function () {
+        progressEl.style.animation = "none";
+        progressEl.style.width = "100%";
+      });
```

Comment:
mobile users can tap to keep a toast visible?

Reply:
Yes, it works, I tested it on my iPad. My code just adds a progress bar and manages its animation. The bootstrap toast has their events to prevent toast from hiding.